src/main/java/com/cleanroommc/groovyscript/compat/mods/advancedmortars/Mortar.java

```diff
@@ -69,16 +75,73 @@ public void remove(RecipeMortar recipe) {
         addBackup(recipe);
     }
 
-    public static List<EnumMortarType> getTypes(RecipeMortar value) {
-        List<EnumMortarType> list = new ArrayList<>();
-        ((RegistryRecipeMortarAccessor) MortarAPI.RECIPE_REGISTRY)
+    public static List<EnumMortarType> getTypes(RecipeMortar recipe) {
+        return ((RegistryRecipeMortarAccessor) MortarAPI.RECIPE_REGISTRY)
                 .getRecipeMap()
-                .forEach((key, value1) -> value1
-                        .stream()
-                        .filter(recipe -> recipe == value)
-                        .map(recipe -> key)
-                        .forEach(list::add));
-        return list;
+                .entrySet()
+                .stream()
+                .filter(x -> x.getValue().contains(recipe))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+    }
+
+    public class RecipeBuilder extends AbstractRecipeBuilder<RecipeMortar> {
+
+        private final List<String> types = new ArrayList<>();
+        private int duration;
+        private ItemStack secondaryOutput;
+        private float secondaryOutputChance;
+
+        public RecipeBuilder type(String type) {
+            this.types.add(type);
+            return this;
+        }
+
+        public RecipeBuilder duration(int duration) {
+            this.duration = duration;
+            return this;
+        }
+
+        public RecipeBuilder secondaryOutput(ItemStack itemStack) {
+            this.secondaryOutput = itemStack;
+            return this;
+        }
+
+        public RecipeBuilder secondaryOutput(ItemStack itemStack, float chance) {
+            this.secondaryOutput = itemStack;
+            this.secondaryOutputChance = chance;
+            return this;
+        }
+
+        @Override
+        public String getErrorMsg() {
+            return "Error adding Advanced Mortars recipe";
+        }
+
+        @Override
+        public void validate(GroovyLog.Msg msg) {
+            validateItems(msg, 1, 8, 1, 1);
+            for (String type : types) {
+                EnumMortarType enumMortarType = EnumMortarType.fromName(type);
+                if (enumMortarType == null) {
+                    msg.add("invalid mortar type: " + type).add("valid types are: " + Arrays.toString(EnumMortarType.NAMES));
+                }
+            }
+            if (secondaryOutput == null) secondaryOutput = ItemStack.EMPTY;
+            if (secondaryOutput.isEmpty()) {
+                secondaryOutputChance = 0.0f;
+            } else if (secondaryOutputChance <= 0.0f) {
```

Comment:
`secondaryOutputChance > 1` should also be validated against

Reply:
The default value for `secondaryOutputChance` is `1.0f`, so this is not needed anymore